---
title: "Supplementary Figures 4 Panels"
geometry: "margin=0.25in"
header-includes:
  - \usepackage{helvet} # changes font to helvetica
  - \usepackage[T1]{fontenc} # changes font to helvetica
  - \renewcommand\familydefault{\sfdefault} # changes font to helvetica
  - \pagenumbering{gobble} # to remove page numbering
  - \thispagestyle{plain} # to remove header
  - \usepackage{setspace}\singlespacing
output:
  pdf_document: default
  html_document:
    df_print: paged
---

Genetic perturbation of PD-1 expression does not alter formation of exhausted CD8+ T cell subsets. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      message = FALSE,
                      warning = FALSE, 
                      fig.align = "center",
                      fig.height = 3, 
                      fig.width = 3*1.618)

library(here)
library(DESeq2)
source(here("helper.R"))
detach("package:SeuratObject", unload = T)
source(here("aesthetics.R"))

```

```{r read in data}
so <- readRDS(here("scRNAseq", "data", "processed_data_objects", "00_seurat_workflow.so.rds"))
umap_coords <- as_tibble(so@reductions$umap@cell.embeddings, rownames = "cell")
metadata <- as_tibble(so@meta.data, rownames = "cell") %>% 
  left_join(umap_coords, by = "cell") 

pca_loadings <- readr::read_tsv(here("scRNAseq", "data", "processed_data_tables", 
                                     "05_pseudobulk.pca_loadings.tsv"))
# flip coordinates for effector-like to be consistent
pca_loadings <- pca_loadings %>% 
  mutate(PC1 = ifelse(cluster_id == "effector-like", -PC1, PC1))

norm_counts <- read_tsv(here("scRNAseq", "data", "processed_data_tables",
                             "05_pseudobulk.norm_counts.tsv"))

gsea_df <- read_tsv(here("scRNAseq", "data", "processed_data_tables",
                         "03_gsea.genotype_one-v-one_per-cluster_log2FC.tsv"))

cluster_dge_df <- read_tsv(here("scRNAseq", "data", "processed_data_tables", 
                                "01_dge.cluster_one-v-all.tsv"))
dge_df <- read_tsv(here("scRNAseq", "data", "processed_data_tables", 
                        "01_dge.genotype_one-v-one_per-cluster.tsv"))
gene_sig_df <- read_tsv(here("scRNAseq", "data", "gene_signatures", 
                             "02_gene_signatures.msigdbr_H-C2-C7.tsv"))

pb_pca_genes <- read_tsv(here("scRNAseq", "data", "gene_signatures", "pb_pca_genes.tsv"))
```

# Supplementary Figure 4a

```{r mean_scaled_norm_counts}
scaled_norm_counts <- norm_counts %>% 
  pivot_longer(cols = 2:ncol(.), 
               names_to = "sample_genotype_cluster", 
               values_to = "norm_counts") %>% 
  separate(sample_genotype_cluster, into = c("sample", "genotype", "cluster"), 
           sep = "_") %>% 
  group_by(gene, cluster) %>% 
  mutate(scaled_norm_counts = as.vector(scale(norm_counts, center = FALSE))) %>% 
  mutate(scaled_norm_counts = ifelse(is.na(scaled_norm_counts), 0, 
                                     scaled_norm_counts)) 
mean_scaled_norm_counts <- scaled_norm_counts %>% 
  group_by(genotype, cluster, gene) %>% 
  summarise(avg_scaled_norm_counts = mean(scaled_norm_counts)) %>% 
  pivot_wider(id_cols = gene, 
              names_from = c(genotype, cluster), 
              values_from = avg_scaled_norm_counts)
mean_scaled_norm_counts_mtx <- mean_scaled_norm_counts %>% 
  column_to_rownames("gene") %>% 
  as.matrix()
```

```{r h_list}
h_list <- list()

col_fun <- colorRamp2(breaks = c(0.8, 1, 1.2), colors = c("white", "darkblue", "darkblue"))
for (cluster_str in names(palette_cluster)) {
  if (cluster_str == "dividing")
    next
  pc1_genes <- pca_loadings %>%
    filter(cluster_id == cluster_str) %>% 
    arrange(desc(PC1)) %>% 
    .$gene
  h_mtx <- mean_scaled_norm_counts_mtx[pc1_genes,paste0(c("W", "D", "K"), "_", cluster_str)]
  colnames(h_mtx) <- c("WT", "EnhDel", "PD-1 KO")
  
  selected_genes <- pb_pca_genes %>% 
    filter(PC == "PC1") %>% 
    filter(Clusters == cluster_str) %>% 
    .$gene
  selected_genes <- selected_genes[selected_genes %in% pc1_genes]
  ha <- rowAnnotation(
    gene = anno_mark(
      at = map_int(selected_genes, ~which(pc1_genes == ..1)), 
      labels = selected_genes, 
      link_gp = gpar(lwd = LINE_WIDTH), 
      labels_gp = gpar(fontsize = TEXT_SIZE), 
      padding = unit(5, "mm")
    )
  )
  
  h_list[[cluster_str]] <- Heatmap(
    h_mtx, 
    width = unit(TEXT_SIZE*9, "pt"), 
    column_title = cluster_abbr_key[cluster_str], 
    column_title_gp = gpar(fontsize = TEXT_SIZE), 
    col = col_fun, 
    name = "mean scaled norm counts", 
    cluster_rows = FALSE, 
    cluster_columns = FALSE, 
    column_names_gp = gpar(fontsize = TEXT_SIZE),
    show_row_names = FALSE, 
    right_annotation = ha
  )
}
```

```{r loading arrow}
loading_arrow <- ggplot() + 
  geom_segment(aes(x = 0, xend = 0, y = -3, yend = 3), 
               arrow = arrow(angle = , 
                             length = unit(TEXT_SIZE / 2, "pt"), 
                             type = "closed"), 
               size = LINE_WIDTH) + 
  annotate("text", label = "PC1 loading", 
           x = -0.5, y = 0, angle = 90,
           size = TEXT_SIZE*GGPLOT_TEXT_SCALE_FACTOR) + 
  ylim(-4.5, 4.5) + xlim(-1, 1) + 
  theme_void()
```

```{r plot_fig4b, fig.width = 7}
plot_supp4a <- function() {
  h_list_draw <- h_list %>% 
    map(~..1 %>% 
          draw(show_heatmap_legend = FALSE) %>% 
          grid.grabExpr()) %>% 
    plot_grid(plotlist = ., ncol = 4)
  lgd <- Legend(col_fun = col_fun, title = "expr.", 
                title_gp = gpar(fontsize = TEXT_SIZE, fontface = "bold"), 
                grid_width = LEGEND_SIZE,
                labels_gp = gpar(fontsize = TEXT_SIZE)) %>% 
    draw() %>% grid.grabExpr()
  plot_grid(loading_arrow, h_list_draw, lgd, 
            ncol = 3, rel_widths = c(1, 8, 1))
}
plot_supp4a()
```

# Supplementary Figure 4b

```{r p_Car2 and p_Cd7}
gene_df <- as_tibble(t(so@assays$RNA@data[c("Car2", "Cd7"), ]), 
                     rownames = "cell")
plot_df <- metadata %>% 
  left_join(gene_df, by = "cell") %>% 
  mutate(Clusters_short = cluster_abbr_key[Clusters]) %>% 
  mutate(Clusters_short = factor(Clusters_short, levels = cluster_abbr_key)) 
plot_gene_by_cluster <- function(plot_df, gene_sym) {
  gene_sym <- enquo(gene_sym)
  
  ann_df <- cluster_dge_df %>% 
    mutate(Clusters_short = cluster_abbr_key[cluster]) %>% 
    mutate(Clusters_short = factor(Clusters_short, levels = cluster_abbr_key)) %>% 
    filter(gene == rlang::as_name(gene_sym) & p_val_adj < 0.05) %>% 
    mutate(y = max(plot_df[[rlang::as_name(gene_sym)]]) + 0.05)

  plot_df %>% 
    ggplot() + 
    aes(Clusters_short, !!gene_sym) + 
    geom_violin(aes(fill = Clusters_short), scale = "width", 
                size = LINE_WIDTH) + 
    stat_compare_means(comparisons = unique(plot_df$Clusters_short)[2:5] %>% 
                                            as.character() %>% 
                                            map(~c(..1, "term.")), 
                       label = "p.signif") + 
    scale_fill_manual(values = palette_cluster_abbr) + 
    labs(y = "Expression Level") + 
    theme(legend.position = "none", 
          axis.title.x = element_blank()) + 
    ggtitle(bquote(~italic(.(rlang::as_name(gene_sym))))) 
}
p_Car2 <- plot_gene_by_cluster(plot_df, Car2)
p_Cd7 <- plot_gene_by_cluster(plot_df, Cd7)
```

```{r plot_supp4b}
plot_supp4b <- function() {
  plot_grid(p_Car2, p_Cd7)
}
plot_supp4b()
```

# Supplementary Figure 4c

```{r gene_df and plot_df}
gene_df <- t(so@assays$RNA@data[c("Tox", "Ikzf2"), ]) %>% 
  as_tibble(rownames = "cell")
plot_df <- metadata %>% 
  left_join(gene_df, by = "cell") %>% 
  mutate(Clusters_short = cluster_abbr_key[Clusters]) %>% 
  mutate(Clusters_short = factor(Clusters_short, levels = cluster_abbr_key)) %>% 
  mutate(genotype = factor(genotype, levels = c("W", "D", "K")))
```

```{r plot_gene_by_genotype}
plot_gene_by_genotype <- function(plot_df, gene_sym) {
  gene_sym <- enquo(gene_sym)
  
  ann_df <- dge_df %>% 
    mutate(Clusters_short = cluster_abbr_key[Clusters]) %>% 
    mutate(Clusters_short = factor(Clusters_short, levels = cluster_abbr_key)) %>% 
    filter(gene == rlang::as_name(gene_sym) & p_val_adj < 0.05) %>% 
    mutate(genotype1_xshift = case_when(
      contrast %>% startsWith("D") ~ -0.33, 
      contrast %>% startsWith("W") ~ 0, 
      contrast %>% startsWith("K") ~ 0.33
    )) %>% 
    mutate(genotype2_xshift = case_when(
      contrast %>% endsWith("D") ~ -0.33, 
      contrast %>% endsWith("W") ~ 0, 
      contrast %>% endsWith("K") ~ 0.33
    )) %>% 
    mutate(x = as.numeric(Clusters_short) + genotype1_xshift, 
           xend = as.numeric(Clusters_short) + genotype2_xshift, 
           y = case_when(
             contrast == "D_K" ~ 3.75, 
             contrast == "K_W" ~ 4.25, 
             contrast == "D_W" ~ 4.75 
           )) %>% 
    mutate(sig = case_when(
      p_val_adj < 0.001 ~ "***", 
      p_val_adj < 0.01 ~ "**", 
      p_val_adj < 0.05 ~ "*"
    )) %>% 
    mutate(sig_x = (x + xend) / 2)
  
  plot_df %>% 
    ggplot() + 
    aes(Clusters_short, !!gene_sym) + 
    geom_violin(aes(fill = genotype), scale = "width", 
                size = LINE_WIDTH) + 
    scale_fill_manual(values = palette_genotype) + 
    labs(y = "Expression Level") + 
    geom_segment(aes(x = x, xend = xend, y = y, yend = y), 
                 data = ann_df) + 
    geom_text(aes(x = sig_x, y = y + 0.05, label = sig), 
              data = ann_df) + 
    theme(legend.position = "none", 
          axis.title.x = element_blank()) + 
    ggtitle(bquote(~italic(.(rlang::as_name(gene_sym))))) 
}
```

```{r p_Tox and p_Ikzf2}
p_Tox <- plot_gene_by_genotype(plot_df, Tox)
p_Ikzf2 <- plot_gene_by_genotype(plot_df, Ikzf2)
```

```{r plot_supp4c, fig.width = 7.5, fig.height = 2.5}
plot_supp4c <- function() {
  plot_grid_shared_legend(plotlist = list(p_Tox, p_Ikzf2), 
                          legend.position = "right", 
                          rel_widths = c(9, 1))
}
plot_supp4c()
```

# Supplementary Figure 4d and 4e

```{r rank_vec_list}
rank_vec_list <- dge_df %>% 
  mutate(Clusters = factor(Clusters, levels = names(palette_cluster))) %>% 
  arrange(Clusters) %>% 
  filter(Clusters != "dividing") %>% 
  filter(contrast != "K_W") %>% 
  named_group_split(contrast, Clusters) %>% 
  map(~{set_names(.$avg_logFC, .$gene)})

apoptosis_genes <- gene_sig_df %>% 
  filter(gs_name == "HAMAI_APOPTOSIS_VIA_TRAIL_UP") %>% 
  .$gene_symbol
apoptosis_leg_list <- rank_vec_list %>% 
  imap(~get_leading_edge_df(..1, apoptosis_genes) %>% 
         mutate(Clusters_contrast = ..2) %>% 
         separate(Clusters_contrast, into = c("contrast", "Clusters"), sep = " / "))
apoptosis_df <- gsea_df %>% 
  filter(gs_name == "HAMAI_APOPTOSIS_VIA_TRAIL_UP") 

treg_genes <- gene_sig_df %>% 
  filter(gs_name == "GSE14350_TREG_VS_TEFF_UP") %>% 
  .$gene_symbol
treg_leg_list <- rank_vec_list %>% 
  imap(~get_leading_edge_df(..1, treg_genes) %>% 
         mutate(Clusters_contrast = ..2) %>% 
         separate(Clusters_contrast, into = c("contrast", "Clusters"), sep = " / "))
treg_df <- gsea_df %>% 
  filter(gs_name == "GSE14350_TREG_VS_TEFF_UP") 
```

```{r cluster and contrast ann}
cluster_ann <- tibble(Clusters = factor(names(palette_cluster_abbr), 
                                        levels = names(palette_cluster_abbr))) %>% 
  filter(Clusters != "div.") %>% 
  ggplot() + 
  aes(Clusters, 0, label = Clusters) + 
  geom_text(size = TEXT_SIZE*GGPLOT_TEXT_SCALE_FACTOR) + 
  theme_void()
contrast_ann <- tibble(contrast = c("EnhDel\nvs.\nPD-1 KO", "EnhDel\nvs.\nWT")) %>% 
  ggplot() + 
  aes(0, contrast, label = contrast) + 
  geom_text(size = TEXT_SIZE*GGPLOT_TEXT_SCALE_FACTOR) + 
  theme_void()
```

```{r reset theme}
theme_set(figure_layout_theme())
```

```{r plot_supp4d, fig.width = 7, fig.asp = 1/3}
plot_supp4d <- function(rel_plot_size = 2, 
                       rel_heights = c(1, 9), 
                       rel_widths = c(1, 12)) {
  p <- plot_gsea_curve_grid(apoptosis_leg_list, apoptosis_df, 
                            rel_plot_size = rel_plot_size, 
                            annotate_coords = c(x = 0.025, y = 0.1), 
                            text_size = 5*GGPLOT_TEXT_SCALE_FACTOR)
  plot_grid(
    NULL, cluster_ann + 
      annotate("text", x = 0.5, y = 1, label = "HAMAI_APOPTOSIS_VIA_TRAIL_UP", size = PLOT_TITLE_SIZE*GGPLOT_TEXT_SCALE_FACTOR, hjust = 0) + 
      ylim(-0.5, 1.5) + 
      scale_x_discrete(expand = c(0, 0.5)), 
    contrast_ann, p, 
    ncol = 2, rel_heights = rel_heights, rel_widths = rel_widths
  )
}
plot_supp4d(rel_plot_size = 1.75)
```

```{r plot_supp4e, fig.width = 7, fig.asp = 1/3}
plot_supp4e <- function(rel_plot_size = 2, 
                       rel_heights = c(1, 9), 
                       rel_widths = c(1, 12)) {
  p <- plot_gsea_curve_grid(treg_leg_list, treg_df, 
                            rel_plot_size = rel_plot_size, 
                            annotate_coords = c(x = 0.025, y = 0.2))
  plot_grid(
    NULL, cluster_ann + 
      annotate("text", x = 0.5, y = 1, label = "TREG_VS_TEFF_UP", size = PLOT_TITLE_SIZE*GGPLOT_TEXT_SCALE_FACTOR, hjust = 0) + 
      ylim(-0.5, 1.5) + 
      scale_x_discrete(expand = c(0, 0.5)), 
    contrast_ann, p, 
    ncol = 2, rel_heights = rel_heights, rel_widths = rel_widths
  )
}
plot_supp4e(rel_plot_size = 1.75)
```

# Supplementary Figure 4 Layout

\newpage

```{r Supplementary Figure 4, fig.height = 10, fig.width = 7.5}
theme_set(figure_layout_theme())

r1_height <- 0.25
r1_bot_align <- 1 - r1_height
r2_height <- 0.15
r2_bot_align <- r1_bot_align - r2_height
r3_height <- 0.25
r3_bot_align <- r2_bot_align - r3_height - 0.025
r4_height <- 0.25
r4_bot_align <- r3_bot_align - r4_height - 0.025

ggdraw() + 
  draw_plot(plot_supp4a(), 0, r1_bot_align, 0.6, r1_height) + 
  draw_plot(plot_supp4b(), 0.6, r1_bot_align + 0.05, 0.4, r1_height - 0.1) +
  draw_plot(plot_supp4c(), 0.15, r2_bot_align, 0.7, r2_height) +
  draw_plot(plot_supp4d(), 0, r3_bot_align, 1, r3_height) +
  draw_plot(plot_supp4e(), 0, r4_bot_align, 1, r4_height) +
  draw_plot_label(label = c("a", "b", "c", "d", "e"), 
                  size = PLOT_TITLE_SIZE, 
                  x = c(0, 0.6, 0, 0, 0), 
                  y = c(r1_height + r1_bot_align, 
                        r1_height + r1_bot_align, 
                        r2_height + r2_bot_align, 
                        r3_height + r3_bot_align,
                        r4_height + r4_bot_align))
```

\newpage

```{r sessionInfo}
sessionInfo()
```
